project: Add last job status as for project sync feedback feature.

seiwailai opened the pull request:
##### SUMMARY
Fixes #9872 
<!--- Describe the change, including rationale and design decisions -->
Add `last job status` as for the project sync feedback feature. When users clicked the `sync` button on the project page, `last job status` will eventually update the status to `Pending`, `Waiting`, `Running`, and final result (`Successful`, `Cancelled`, `Failed`, `Error`). The implementation requires the WebSocket connection to ensure synchronous status update.

In particular, the `last job status` is similar to the functionality of `status` in the project list (`status` in `ProjectListItem.jsx`). More specifically, the `last job status` accompanied by a job link that allowing the user to navigate to the job output page. Besides, there is also the tooltip that allowing the user to view the information related to the most recent sync, covering information like `JOB ID`, `STATUS`, `FINISHED`.

The rationale of having the `last job status` instead of redirection or toast notification is that:
1. User has choices on whether to redirect to the job output. If the user wishes to navigate to job output, he/she can click the link. Besides, the user might have other projects to be synced right after the current project and he/she may want to proceed back to the project list page instead of the job output page. If we implement force redirection, it would take a longer time to navigate to the project list page.
2. The status update on `last job status` is fundamentally similar to toast notification where the user can immediately be notified if he/she already clicked the `sync` button to launch the job.

Nevertheless, this PR requires further discussion. Any comments are welcomed!

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - UI
 
1. `awx/ui_next/src/screens/Project/Project.jsx`
2. `awx/ui_next/src/screens/Project/useWsProject.js` - Added websocket implementation
3. `awx/ui_next/src/screens/Project/ProjectDetail/ProjectDetail.jsx`

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 19.0.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
In case if users spam the sync button, we will need to ensure the fluent UI on the most recent sync tooltip and last job status. Thus, we would not want to update our last job status to `Pending` if there is a current running job.

For instance, we clicked sync for a particular project twice. 
1. For the first sync, our last job status should immediately change to `Pending`, then `Waiting`, then `Running`, then result (which are `Successful`, `Failed`, `Error`, `Cancelled`).
2. For the second sync, if we have a running job, we should not update our UI to `Pending`, otherwise our most recent sync tooltip UI will lose our current running job and we cannot navigate to the job link through the link provided by `last job status` tooltip.

**Issue of `sync` button click spam**
Ideally, we should prevent any spamming on the `sync` button using backend logic to reduce overload on the server as we already have a similar running project. Together with backend logic, we can disable the sync button right after we start to sync a project.

However, if we only disable `sync` through the frontend, this seems insecure as people with bad intentions might able to change the button disable attribute.

<!--- Paste screenshot/gif below, e.g. before and after your change -->
**After**
![project_sync_feedback](https://user-images.githubusercontent.com/67208821/115053818-1f66f180-9f12-11eb-9970-55adfce6abf9.gif)

seiwailai committed:
project: Add last job status as for feedback feature.

nixocio commented:
Thanks for your PR. Is it ready to be reviewed?

seiwailai commented:
> Thanks for your PR. Is it ready to be reviewed?

Yup, this is ready for review

AlexSCorey commented on awx/ui_next/src/screens/Project/useWsProject.js:
```diff
@@ -0,0 +1,68 @@
+      // to `Pending`, then `Waiting`, then `Running`, then result
+      // (which are `successful`, `failed`, `error`, `cancelled`.
+      // For second sync, if the status response is `pending` and we have
+      // running and waiting jobs, we should not update our UI to `Pending`,
```
```suggestion
      // running or waiting jobs, we should not update our UI to `Pending`,
```

AlexSCorey replied:
We should also disable the sync button with a tooltip to explain why it is cancelled

seiwailai replied:
> I really like this! Just a couple of minor changes. We should probably add some tests as well.

@AlexSCorey Thanks for your review, will make changes and add tests as soon as possible!

AlexSCorey reviewed (changes requested):
I really like this!  Just a couple of minor changes.  We should probably add some tests as well.

jakemcdermott commented:
recheck

mabashian commented:
@AlexSCorey is this just waiting for a re-review?

seiwailai committed:
project: Add disable sync button feature.

Disable sync button if there is any pending, waiting or running job

seiwailai committed:
useWsProject: Add tests for useWsProject.jsx.

Add tests like 'should return project detail', 'should establish websocket connection' and 'should update project status'.

seiwailai committed:
ProjectSyncButton: Add tests.

Add 'disable button and set onClick to undefined on sync' and 'should render tooltip on sync'

seiwailai commented:
@AlexSCorey @nixocio @jakemcdermott This is ready for review!

AlexSCorey reviewed (approved):
Good well done!  Thanks for doing this.

tiagodread reviewed (changes requested):
Hey @seiwailai thanks for adding all this context, this is awesome!
We have the same issue with the sync button span on the Project list, do you mind disabling the sync button on Project list as well?

seiwailai commented:
> Hey @seiwailai thanks for adding all this context, this is awesome!
> We have the same issue with the sync button span on the Project list, do you mind disabling the sync button on Project list as well?

@tiagodread Sure thing!

seiwailai committed:
ProjectListItem: Disabled sync button and render tooltip on sync.

Added feature of disabling sync button when users click sync button and rendering tooltip when disabled sync button on hover.

seiwailai commented:
> Hey @seiwailai thanks for adding all this context, this is awesome!
> We have the same issue with the sync button span on the Project list, do you mind disabling the sync button on Project list as well?

@tiagodread This is after changes for the project list!

![project_list_sync_button](https://user-images.githubusercontent.com/67208821/116518675-24a94080-a903-11eb-94e4-694dfe1712a3.gif)

tiagodread commented:
@seiwailai trying to create a project when I hit the save button the page is crashing:

![chrome-capture (9)](https://user-images.githubusercontent.com/12464858/116559708-17765c80-a8d7-11eb-9053-67ee1257024d.gif)
![image](https://user-images.githubusercontent.com/12464858/116559702-16ddc600-a8d7-11eb-866b-ea05e8539e3b.png)

seiwailai committed:
Project: Added project last job status UI with websocket feature

1. Activate web socket once get into project detail page to ensure job status update synchronization.\n 2. Show last job status if there is no current job.\n 3. Show current job status if there is any current pending, waiting or running job.

seiwailai committed:
ProjectList: Change job status UI update logic.

1. Render current job status if there is current waiting, pending or running job.\n 2. Render last job status if there is no current job.

seiwailai commented:
@tiagodread I've added the conditional statement on rendering job status UI to solve the crash issue after project creation. Thank you for your comment!

**Newly-created project on sync**
![add_project_on_sync](https://user-images.githubusercontent.com/67208821/116746348-ef9a0c80-aa2e-11eb-8950-beda4d828815.gif)

**Project on sync**
![project_on_sync](https://user-images.githubusercontent.com/67208821/116746380-fa54a180-aa2e-11eb-839e-4472a9bad2c0.gif)

**Project list on sync**
![project_list_on_sync](https://user-images.githubusercontent.com/67208821/116746402-03de0980-aa2f-11eb-9c73-62c519432000.gif)

tiagodread reviewed (approved):
Verified that:
- The user can create projects
- The sync button is disabled while the sync is running and the tooltip is displayed
- The user can click on the `Last Job Status` and navigate to the `Source Control Update` for the current project.
- All e2e tests are 🟢
